Tensile/TensileCreateLibrary.py

```diff
@@ -103,31 +103,109 @@ def processKernelSource(kernel, kernelWriterSource, kernelWriterAssembly) -> Pro
+def linkIntoCodeObject(
+    objFiles: List[str], coPathDest: Union[Path, str], kernelWriterAssembly: KernelWriterAssembly
+):
+    """Links object files into a code object file.
+
+    Args:
+        objectFiles: A list of object files to be linked.
+        coPathDest: The destination path for the code object file.
+        kernelWriterAssembly: An instance of KernelWriterAssembly to get link arguments.
+
+    Raises:
+        RuntimeError: If linker invocation fails.
+    """
+    args = []
+    if os.name == "nt":
+        # On Windows, the objectFiles list command line (including spaces)
+        # exceeds the limit of 8191 characters, so using response file
+
+        responseFile = os.path.join("/tmp", "clangArgs.txt")
+        with open(responseFile, "wt") as file:
+            file.write(" ".join(objFiles))
+            file.flush()
+
+        args = kernelWriterAssembly.getLinkCodeObjectArgs(["@clangArgs.txt"], str(coPathDest))
+    else:
+        args = kernelWriterAssembly.getLinkCodeObjectArgs(objFiles, str(coPathDest))
+
+    tPrint(2, "Linking objects into co files: " + " ".join(args))
+    try:
+        out = subprocess.check_output(args, stderr=subprocess.STDOUT)
+        tPrint(2, f"Output: {out}")
+    except subprocess.CalledProcessError as err:
+        raise RuntimeError(
+            f"Error linking object files in to code object: {err.output}\nFailed command: {' '.join(args)}"
+        )
+
+
+def compressCodeObject(
+    coPathSrc: Union[Path, str], coPathDest: Union[Path, str], gfx: str, bundler: str
+):
+    """Compresses a code object file using the provided bundler.
+
+    Args:
+        coPathSrc: The source path of the code object file to be compressed.
+        coPathDest: The destination path for the compressed code object file.
+        gfx: The target GPU architecture.
+        bundler: The path to the Clang Offload Bundler executable.
+
+    Raises:
+        RuntimeError: If compressing the code object file fails.
+    """
+    args = [
+        bundler,
+        "--compress",
+        "--type=o",
+        "--bundle-align=4096",
+        f"--targets=host-x86_64-unknown-linux,hipv4-amdgcn-amd-amdhsa--{gfx}",
+        "--input=/dev/null",
+        f"--input={str(coPathSrc)}",
+        f"--output={str(coPathDest)}",
+    ]
+
+    tPrint(2, f"Bundling/compressing code objects: {' '.join(args)}")
+    try:
+        out = subprocess.check_output(args, stderr=subprocess.STDOUT)
+        tPrint(2, f"Output: {out}")
+    except subprocess.CalledProcessError as err:
+        raise RuntimeError(
+            f"Error compressing code object via bundling: {err.output}\nFailed command: {' '.join(args)}"
+        )
+
+
+def buildAssemblyCodeObjectFiles(kernels, kernelWriterAssembly, outputPath, removeTemporaries):
+
+    countAsmKernels = lambda kernels: sum(k["KernelLanguage"] == "Assembly" for k in kernels)
+
+    extObj = ".o"
+    extCo = ".co"
+    extCoRaw = ".co.raw"
```

Comment:
what is the meaning of raw?

Reply:
Here, ".raw" is an extension used to identify the co file after linking but before compression. Files with this extension live in the *build_tmp/* directory and are the input for the compression via bundling step.